### PR TITLE
LPD-5118 Create feature flag fixture

### PR DESCRIPTION
**Pull request by izaera**

### CODE REVIEW

**Comment by izaera** [comment]
ci:test:playwright

File: modules/apps/feature-flag/feature-flag-web/src/main/java/com/liferay/feature/flag/web/internal/jaxrs/application/FeatureFlagApplication.java [hunk 1]
```diff
@@ -80,6 +80,32 @@ public Set<Object> getSingletons() {
...
+		FeatureFlagsBag featureFlagsBag =
+			_featureFlagsBagProvider.getOrCreateFeatureFlagsBag(companyId);
+
+		FeatureFlag featureFlag = featureFlagsBag.getFeatureFlag(key);
+
+		return Response.ok(
+			HashMapBuilder.<String, Object>put(
+				"dependentFeatureFlags",
+				TransformUtil.transform(
+					_getDependentFeatureFlags(featureFlagsBag, key),
```
Comment (dsanz) [inline modules/apps/feature-flag/feature-flag-web/src/main/java/com/liferay/feature/flag/web/internal/jaxrs/application/FeatureFlagApplication.java:99]: does this return the transitive dependencies? (I'm not even sure they can exist, just confirming we're all ok with the model)

File: modules/test/playwright/tests/batch-planner/import.spec.ts [hunk 2]
```diff
@@ -359,17 +361,13 @@ test('can import CSV file with an unexisting field', async ({
 	]);
 
 	await apiHelpers.objectAdmin.deleteObjectDefinition(response.id);
-	await apiHelpers.featureFlag.updateFeatureFlag('COMMERCE-8087', false);
-	await apiHelpers.featureFlag.updateFeatureFlag('LPS-173135', false);
```
Comment (dsanz) [inline modules/test/playwright/tests/batch-planner/import.spec.ts:363]: Here test is turning 2 FF off however in the beginning it's enabling only one. I don't see any dependency across them in portal.properties, so to me this is a bit of a mystery

File: modules/apps/feature-flag/feature-flag-web/src/main/java/com/liferay/feature/flag/web/internal/jaxrs/application/FeatureFlagApplication.java (hunk 1 above)
Comment (izaera) [reply modules/apps/feature-flag/feature-flag-web/src/main/java/com/liferay/feature/flag/web/internal/jaxrs/application/FeatureFlagApplication.java:99]: Yep, that's the theory. That code was introduced to refresh the FF configuration screen after a modification without needing to reload the page.

I've been unable to check it myself, though. :grimacing:

File: modules/test/playwright/tests/batch-planner/import.spec.ts (hunk 2 above)
Comment (izaera) [reply modules/test/playwright/tests/batch-planner/import.spec.ts:363]: Oh yes, I forgot to mention it: the `LPS-173135` comes from and old poshi test that was migrated to playwright. However that FF doesn't exist any more so I've simply ignored it.

My guess is that it was a dependency FF that needed to be disabled after the test because there was no support for dependent FFs in poshi (but take it as a hypothesis).

File: modules/test/playwright/fixtures/featureFlagsTest.ts [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+	return result;
+}
+
+async function invokeServer<T>(
+	page: Page,
+	url: string,
+	partialBody: object
+): Promise<T> {
+	return await page.evaluate(
+		async ({partialBody, url}) =>
```
[the remark this answers is not in the document]
Comment (izaera) [reply modules/test/playwright/fixtures/featureFlagsTest.ts:150]: It should, since playwright controls it and `page.evaluate()` is already bypassing all security mechanisms (since it's executing code from the test in the context of the page). But we will see...

File: modules/test/playwright/fixtures/featureFlagsTest.ts [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+	key: string;
+	title: string;
+}
+
+async function isEnabled(key: string, page: Page): Promise<IsEnabledResult> {
+	return invokeServerWithRetry(
+		page,
+		'/o/com-liferay-feature-flag-web/is-enabled',
+		{key}
+	);
```
Comment (markocikos) [inline modules/test/playwright/fixtures/featureFlagsTest.ts:110]: Can we use `Liferay.FeatureFlags['LPS-*']` here? I realize it is not checking the actual state of FF in backend, for example changed by `setEnabled`, but shouldn't UI be reloaded after FF change in any case?

File: modules/test/playwright/fixtures/featureFlagsTest.ts [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+
+export type FeatureFlags = FeatureFlag[];
+
+export interface FeatureFlag {
+	readonly enabled: boolean;
+	readonly key: string;
+}
+
+function featureFlagsTest(options: FeatureFlagsOptions) {
+	const fixtureImpl = test.extend<{
```
Comment (markocikos) [inline modules/test/playwright/fixtures/featureFlagsTest.ts:22]: Sneaky? I am not sure if fixtures were really intended to be defined this way, as plain old functions, but it doesn't really matter. It allows for a very nice API to set FFs :+1:

File: modules/test/playwright/fixtures/featureFlagsTest.ts [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+					originalFeatureFlags.push(...result.dependentFeatureFlags);
+				}
+
+				try {
+
+					// Activate  FFs
+
+					for (const [key, enabled] of Object.entries(options)) {
+						await setEnabled(enabled, key, page);
+					}
```
Comment (markocikos) [inline modules/test/playwright/fixtures/featureFlagsTest.ts:45]: This is disabling FFs as well as activating, right? Comment may be putting me off.

File: modules/test/playwright/fixtures/featureFlagsTest.ts [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+
+					originalFeatureFlags.reverse();
+
+					for (const featureFlag of originalFeatureFlags) {
+						await setEnabled(
+							featureFlag.enabled,
+							featureFlag.key,
+							page
+						);
+					}
```
Comment (markocikos) [inline modules/test/playwright/fixtures/featureFlagsTest.ts:75]: I am not sure if we want to immediately restore original state. What if FF changes backend behavior, and tests depend on the changed behavior? I would assume FF recovery needs to be done only after all tests are completed.

**Review by markocikos — COMMENTED** [review]
@izaera Few questions inline

File: modules/test/playwright/fixtures/featureFlagsTest.ts (hunk 4 above)
Comment (izaera) [reply modules/test/playwright/fixtures/featureFlagsTest.ts:110]: Mmm, good idea, I'll test that approach :+1:

File: modules/test/playwright/fixtures/featureFlagsTest.ts (hunk 5 above)
Comment (izaera) [reply modules/test/playwright/fixtures/featureFlagsTest.ts:22]: I assume it's perfectly valid, since the way the framework is constructed seems to be aligned with this usage. But I agree it's a bit "advanced" :sweat_smile: 

I considered using [fixture options](https://playwright.dev/docs/test-fixtures#fixtures-options) but they are too coarse grained for our needs, I think.

File: modules/test/playwright/fixtures/featureFlagsTest.ts (hunk 6 above)
Comment (izaera) [reply modules/test/playwright/fixtures/featureFlagsTest.ts:45]: True. I created a first version that only enables but then people asked me to add disabling too and I didn't update the comment.

File: modules/test/playwright/fixtures/featureFlagsTest.ts (hunk 7 above)
Comment (izaera) [reply modules/test/playwright/fixtures/featureFlagsTest.ts:75]: > What if FF changes backend behavior, and tests depend on the changed behavior?

Well, if FF changes behavior and it is needed for another test you have to use the fixture from other test. That's the intention of this fixture: eliminating side effects (from one tests to others).

> I would assume FF recovery needs to be done only after all tests are completed.

It is set and restored per test. So, in the same file, if there are 3 tests using it those 3 tests will turn the FF on on entering, then restoring it on exit.

I think the misunderstanding here is that you think the declaration  `featureFlagsTest(...)` is affecting the whole file but that is not true. What `mergeTests` does is composing several fixtures and it's after, when you invoke the returned `test`  method, that the fixture is really used.

In other words, you could have two different mergings in the same file, which clearly demonstrates what I'm tying to convey:


```typescript
const testWithFFa = mergeTests(loginTest, featureFlagsTest({"LPS-A": true}));
const testWithFFb = mergeTests(loginTest, featureFlagsTest({"LPS-B": true}));

testWithFFa('something happens' ...);

testWithFFa('other things happens' ...);
```

Comment (izaera) [reply modules/test/playwright/fixtures/featureFlagsTest.ts:75]: One possible optimization to the current logic would be letting the FFs changed at the end of the tests, then change them only when entering the setup phase again.

However this optimization needs the state of the server to be stored somewhere so that we don't lose track of what it is. For one single worker that's definitely possible if we promote the `originalFeatureFlags` variable to global module scope. 

However, for several workers that setup wouldn't work because we would need to distribute that information among the workers which would be difficult (even impossible).

It's true that with several workers even the current code fails because two workers may collide when setting/restoring FFs, but at least the current code seems more stable (to me). I prefer to start slow (conservative) :shrug:

**Comment by izaera** [comment]
I've pushed some changes and the fixture tsdoc so that it appears in the IDEs and helps people to write tests (hopefully).

**Comment by izaera** [comment]
ci:test:sf

**Comment by izaera** [comment]
ci:test:playwright

File: modules/test/playwright/fixtures/featureFlagsTest.ts (hunk 7 above)
Comment (markocikos) [reply modules/test/playwright/fixtures/featureFlagsTest.ts:107]: No no, I didn't mean about interaction between tests. Assume we have only one test. What I meant is that test may depend on what FF does in backend. FFs don't only toggle parts of UI, they change Java logic, enable endpoints etc. A fixture resolves before test, and the current solution in PR immediately restores it. If I'm understanding this correctly, this means the state in the tests would be UI with overridden FFs, but the backed with unchanged FFs. If tests depend on changed backend logic, they will fail.

**Review by markocikos — APPROVED** [review]
@izaera LGTM! :sparkles:

File: modules/test/playwright/fixtures/featureFlagsTest.ts (hunk 7 above)
Comment (izaera) [reply modules/test/playwright/fixtures/featureFlagsTest.ts:107]: For the record, I talked to @markocikos by DM and the misunderstanding comes from the fact that he wasn't aware that the invocation of `use()` [inside the fixture](https://github.com/liferay-frontend/liferay-portal/pull/3955/files#diff-ab0aab68b4c280e9bdbf15330c8303a7459aec9c465862b2a90f09d3eab5f349R81) is in fact executing the test. 

Knowing that everything makes sense :+1: .

**Comment by izaera** [comment]
ci:forward

**Comment by izaera** [comment]
ci:test:stable

**Comment by izaera** [comment]
ci:test:stable

**Comment by izaera** [comment]
Failures are not related, but I'm trying stable to pass once more.

**Comment by izaera** [comment]
ci:forward:force

**Comment by izaera** [comment]
ci:forward:force
